app/classes/content_filter.rb

```diff
@@ -1,96 +1,117 @@
+  #     Observation search -- add a new pattern_search file and class.
+  #
+  # There are probably other steps/files I've forgotten. JDC 2016-09-15
 
-  def has_images
-    {
-      sym:          :has_images,
-      model:        Observation,
-      checkbox:     :has_images_checkbox,
-      checked_val:  "NOT NULL",
-      off_val:      "off",
-      on_vals:      ["NOT NULL", "NULL"],
-      sql_cond:     "observations.thumb_image_id IS #{params[:has_images]}"
-    }
-  end
+  attr_accessor :sym
+  attr_accessor :model
+  attr_accessor :checkbox
+  attr_accessor :on_vals
+  attr_accessor :checked_val
+  attr_accessor :off_val
 
-  def has_specimen
-    {
-      sym:          :has_specimen,
-      model:        Observation,
-      checkbox:     :has_specimen_checkbox,
-      checked_val:  "TRUE",
-      off_val:      "off",
-      on_vals:      ["TRUE", "FALSE"],
-      sql_cond:     "observations.specimen IS #{params[:has_specimen]}"
-   }
+  def initialize(opts)
+    @sym          = opts[:sym]
+    @model        = opts[:model]
+    @checkbox     = opts[:checkbox]
+    @on_vals      = opts[:on_vals]
+    @checked_val  = opts[:checked_val]
+    @off_val      = opts[:off_val]
   end
 
-  def filters
-    [has_images, has_specimen]
+  # Array of all available filters.
+  def self.all
+    @@filters ||= [
+      ContentFilter.new(
+        sym:          :has_images,
+        model:        Observation,
+        checkbox:     :has_images_checkbox,
+        on_vals:      ["yes", "no"],
+        checked_val:  "yes",
+        off_val:      nil
+      ),
+      ContentFilter.new(
+        sym:          :has_specimen,
+        model:        Observation,
+        checkbox:     :has_specimen_checkbox,
+        on_vals:      ["yes", "no"],
+        checked_val:  "yes",
+        off_val:      nil
+      )
+    ]
   end
 
-  ### These reflect on above methods.
-  def observation_filters
-    filters.select { |fltr| fltr[:model] == Observation }
+  # Lookup a filter by name (symbol).
+  def self.find(key)
+    all.select { |fltr| fltr.sym == key }.first
   end
 
-  # array of observation_filter names as symbols
-  def observation_filter_keys
-    observation_filters.each_with_object([]) { |fltr, keys| keys << fltr[:sym] }
+  # Array of filters which apply to obserations.
+  def self.observation_filters
+    all.select { |fltr| fltr.model == Observation }
   end
 
-  # Lets view check whether to create checkbox.
-  def observation_filters_with_checkboxes
-    observation_filters.select { |fltr| fltr[:checkbox].present? }
+  # Array of observation filter names as symbols.
+  def self.observation_filter_keys
+    observation_filters.each_with_object([]) { |fltr, keys| keys << fltr.sym }
   end
 
-  module_function(:has_images, :has_specimen, :filters, :observation_filters,
-                  :observation_filter_keys)
+  # Array of filters which use checkboxes.
+  def self.observation_filters_with_checkboxes
```

Comment:
Do we have a motivating example for a filter without a checkbox?  I could see this once we get to location or taxon based filters, but it's not clear to me what happens at that point and whether the has_images and has_specimen filters remain special.